### PR TITLE
Separate out object rounding into its own combinator

### LINKED ISSUE #298: Separate rounding into its own combinator
12 of the 30 symbolic object primitives have a roundness parameter --- aggressively suggesting that roundness should be its own orthogonal concern. I'd suggest excising the roundness parameter from every primitive, and introducing a new one:

```haskell
withRounding ::  Object obj vec => Double -> obj -> obj
```

as well as:

```haskell
getRounding :: Object obj vec => obj -> Double
```

subject to the laws:

```haskell
-------------
withRounding r1 . withRounding r2 = withRounding r1
getRounding . withRounding r = const r
withRounding (getRounding x) x = x

-------------
withRounding r . scale x = scale x . withRounding r
withRounding r . translate x = translate x . withRounding r
withRounding r . rotate x = rotate x . withRounding r
withRounding r . complement = complement  -- don't distribute over complement

-------------
withRounding r (union x) = unionR r x
withRounding r (intersect x) = intersectR r x
withRounding r (difference x) = differenceR r x
withRounding r (cube x) = cubeR r x
-- etc for all currently roundable primitives
```

### PR DESCRIPTION
**Pull request by isovector**
This PR introduces a new combinator `withRounding` which sets the object-rounding for any primitive shapes contained inside. By doing so, we can eliminate all of the `*R` primitives. It's implemented by pushing around an "object context" inside of `getImplicit`, which tracks the current rounding value, and can be used by any primitive which needs it.

Also, there are lots of new tests around this functionality.

The scad interface is unchanged.

Fixes #298

### CODE REVIEW

File: Graphics/Implicit/Export/SymbolicFormats.hs [hunk 1]
```diff
@@ -100,24 +100,29 @@ buildShared (Scale v obj) = call "scale" (bf <$> elements v) [build obj]
 
 buildShared (Mirror v obj) = callNaked "mirror" [ "v=" <> bvect v ] [build obj]
 
+-- TODO(sandy): 'r' is not a rounding parameter
```
Comment (julialongtin) [inline Graphics/Implicit/Export/SymbolicFormats.hs:103]: your todo here is a bit unclear?

File: Graphics/Implicit/Export/SymbolicObj2.hs [hunk 2]
```diff
@@ -36,10 +35,11 @@ symbolicGetOrientedContour res symbObj = orient <$> symbolicGetContour res symbO
         orient (Polyline []) = Polyline []
         orient (Polyline [_]) = Polyline []
 
-symbolicGetContour :: ℝ -> SymbolicObj2 -> [Polyline]
-symbolicGetContour _ (SquareR 0 (V2 dx dy)) = [Polyline [V2 0 0, V2 dx 0, V2 dx dy, V2 0 dy, V2 0 0]]
+symbolicGetContour :: ℝ -> ObjectContext ->  SymbolicObj2 -> [Polyline]
+symbolicGetContour _ ctx (Square (V2 dx dy)) | objectRounding ctx == 0 =
```
Comment (julialongtin) [inline Graphics/Implicit/Export/SymbolicObj2.hs:39]: I usually bump guards to the next line.

File: Graphics/Implicit/ObjectUtil/GetImplicit2.hs [hunk 3]
```diff
@@ -10,7 +10,7 @@ module Graphics.Implicit.ObjectUtil.GetImplicit2 (getImplicit2) where
 import Prelude(cycle, (/=), uncurry, fst, Eq, zip, drop, abs, (-), (/), sqrt, (*), (+), length, fmap, (<=), (&&), (>=), (||), odd, ($), (>), filter, (<), minimum, (.), sin, cos)
 
 import Graphics.Implicit.Definitions
-    ( SymbolicObj2(SquareR, Circle, PolygonR, Rotate2, Shared2), SharedObj (Empty), Obj2, ℝ2, ℝ )
+    (objectRounding, ObjectContext,  SymbolicObj2(Square, Circle, Polygon, Rotate2, Shared2), SharedObj (Empty), Obj2, ℝ2, ℝ )
```
Comment (julialongtin) [inline Graphics/Implicit/ObjectUtil/GetImplicit2.hs:13]:
```suggestion
    ( objectRounding, ObjectContext, SymbolicObj2(Square, Circle, Polygon, Rotate2, Shared2), SharedObj (Empty), Obj2, ℝ2, ℝ )
```

File: Graphics/Implicit/ObjectUtil/GetImplicit3.hs [hunk 4]
```diff
@@ -8,7 +8,7 @@ module Graphics.Implicit.ObjectUtil.GetImplicit3 (getImplicit3) where
 import Prelude (Either(Left, Right), abs, (-), (/), (*), sqrt, (+), atan2, max, cos, minimum, ($), sin, pi, (.), Bool(True, False), ceiling, floor, pure, error, (==), otherwise)
 
 import Graphics.Implicit.Definitions
-    ( ℕ, SymbolicObj3(CubeR, Sphere, Cylinder, Rotate3, ExtrudeR, ExtrudeRM, ExtrudeOnEdgeOf, RotateExtrude, Shared3, ExtrudeRotateR), Obj3, ℝ2, ℝ, fromℕtoℝ, toScaleFn )
+    (objectRounding, ObjectContext,  ℕ, SymbolicObj3(Cube, Sphere, Cylinder, Rotate3, Extrude, ExtrudeM, ExtrudeOnEdgeOf, RotateExtrude, Shared3, ExtrudeRotateR), Obj3, ℝ2, ℝ, fromℕtoℝ, toScaleFn )
```
Comment (julialongtin) [inline Graphics/Implicit/ObjectUtil/GetImplicit3.hs:11]:
```suggestion
    ( objectRounding, ObjectContext, ℕ, SymbolicObj3(Cube, Sphere, Cylinder, Rotate3, Extrude, ExtrudeM, ExtrudeOnEdgeOf, RotateExtrude, Shared3, ExtrudeRotateR), Obj3, ℝ2, ℝ, fromℕtoℝ, toScaleFn )
```

File: Graphics/Implicit/ObjectUtil/GetImplicitShared.hs [hunk 5]
```diff
@@ -9,12 +9,12 @@
...
 module Graphics.Implicit.ObjectUtil.GetImplicitShared (getImplicitShared, normalize) where
 
-import {-# SOURCE #-} Graphics.Implicit.Primitives (Object(getImplicit))
+import {-# SOURCE #-} Graphics.Implicit.Primitives (Object(getImplicit'))
 
 import Prelude (flip, (-), (*), (>), (<), (&&), (/), product, abs, (**), fmap, (.), negate, ($), const)
 
 import Graphics.Implicit.Definitions
-    ( SharedObj(Empty, Full, Complement, UnionR, IntersectR, DifferenceR, Translate, Scale, Mirror, Shell, Outset, EmbedBoxedObj), ComponentWiseMultable((⋯/)), ℝ, minℝ )
+    (objectRounding, ObjectContext,  SharedObj(Empty, Full, Complement, UnionR, IntersectR, DifferenceR, Translate, Scale, Mirror, Shell, Outset, EmbedBoxedObj, WithRounding), ComponentWiseMultable((⋯/)), ℝ, minℝ )
```
Comment (julialongtin) [inline Graphics/Implicit/ObjectUtil/GetImplicitShared.hs:17]:
```suggestion
    ( objectRounding, ObjectContext, SharedObj(Empty, Full, Complement, UnionR, IntersectR, DifferenceR, Translate, Scale, Mirror, Shell, Outset, EmbedBoxedObj, WithRounding), ComponentWiseMultable((⋯/)), ℝ, minℝ )
```

**Review by julialongtin — APPROVED** [review]

File: Graphics/Implicit/Export/SymbolicFormats.hs (hunk 1 above)
Comment (isovector) [reply Graphics/Implicit/Export/SymbolicFormats.hs:103]: This function is removed in #364 so I wasn't too worried about it.

**Pull request merged**